lib/http_put.js

```diff
@@ -0,0 +1,340 @@
+            readSize += pushSize;
+
+            // We must switch streams
+            if (readSize === nextBoundary) {
+                ++chunkId;
+                nextBoundary = chunkId === nChunks ?
+                    size : (chunkId + 1) * chunkSize;
+
+                chunkStream.push(null); // terminates current stream
+                chunkStream = new stream.PassThrough();
+
+                // Kick start callback on next chunk
+                chunkCallback(chunkStream, nextBoundary - readSize,
+                              chunkId, callbackArgs);
+            }
+        } while(readSize + leftover.length > nextBoundary);
+    });
+
+    // Propagate error
+    inputStream.on('error', err =>
+                   chunkStream.emit('error', err));
+
+    // Kick-start everything
+    inputStream.resume();
+    chunkCallback(chunkStream, nextBoundary - readSize,
+                  chunkId, callbackArgs);
+     */
+}
+
+/**
+ * Demultiplex a stream for replication
+ *
+ * Helper to move from single output to
+ * multiple identical output streams
+ *
+ * @param {stream.Readable} inputStream - Stream to chunk
+ * @param {Number} size - Total stream size
+ * @param {Number} nReplica - Number of output stream
+ * @param {function} replicaCallback - Invoked on each replica stream
+ *                    (repStream, size, repId, callbackArgs) -> undefined
+ * @param {Object} callbackArgs - Arguments to pass when invoking callback
+ * @return {undefined}
+ */
+function replicationStreamDemux(
+    inputStream, size, nReplica, replicaCallback, callbackArgs) {
+    /* Shortcut for non-replicated use cases */
+    if (nReplica === 1) {
+        replicaCallback(inputStream, size, 0, callbackArgs);
+        return;
+    }
+
+    const replicatedStreams = [...Array(nReplica).keys()]
+              .map(() => new stream.PassThrough());
+
+    /* Propagate error */
+    inputStream.on('error', err =>
+                   replicatedStreams.forEach(
+                       rstream => rstream.emit('error', err)));
+
+    /* Kick start PUTs */
+    replicatedStreams.forEach((rstream, ridx) => {
+        inputStream.pipe(rstream);
+        replicaCallback(rstream, size, ridx, callbackArgs);
+    });
+}
+
+/**
+ * Decide what is the final status of the operation
+ * and what to do on error fragments.
+ *
+ * @param {Object} opContext - Operation context
+ * @return {Error|null} error - what to feed to client callback
+ * @return {Object|null} errorHandler - Object tracking cleanup action
+ */
+function decidePUT(opContext) {
+    /* Any failure fails all
+     * But if too many fragments of a chunk are in limbo,
+     * ie with unknown status, fail */
+    const perChunkThreshold = opContext.fragments.code === 'CP' ?
+              Math.floor(opContext.fragments.nDataParts / 2) :
+              opContext.fragments.nCodingParts;
+
+    const success = opContext.status.every(
+        chunk => (chunk.nOk > perChunkThreshold && chunk.nError === 0));
```

Comment:
No, it is exactly as you said. Erasure coding check is wrong. As for failing if any error occurs, yeah I guess it can be relaxed (tbd with PM). I implemented the specs I have on this one: "If data cannot be written (one or more fragment failed), the whole write should fail as well.".

Reply:
Question raised to PM.